sqs_workers/memory_sqs.py

```diff
@@ -158,19 +160,40 @@ def delete_messages(self, Entries):
         message_ids = {entry["Id"] for entry in Entries}
 
         successfully_deleted = set()
-        push_back_messages = []
 
-        for message in self.messages:
+        for i, message in enumerate(self.in_flight):
```

Comment:
I feel like my gap in understanding this library might be obvious here. I'm missing something.

This means `queue.delete_messages()` will only act on messages that are in-flight. Isn't that a somewhat unexpected behavior, considering the function's name? Same for  `change_message_visibility_batch`.

Reply:
It might sound unintuitive based on the name, but, in fact this is the way SQS is meant to be used! "Deleting" a message in SQS is equivalent to "acknowledging" a message in Kafka if you're more familiar with that -- you can think of it like there are 3 "states" a message in a queue can be in:

1. Unprocessed, waiting for a worker to pick it up (this is represented by `self.messages`)
2. Currently being processed by a worker -- still around, but not available to be picked up (this is "in flight")
3. Finished with processing, all good (this is represented by deleting the message from the queue)

Only in exceptional cases would you need to skip from state 1 to state 3 -- 99% of the time you only delete a message that's "in flight" (once you're done processing it)! And since like I mentioned `MemoryQueue` is pretty much a toy implementation for tests, I figured it was OK to just model the 99%-of-the-time case, WDYT?

(the existing logic really made no sense to me -- it has no concept of "in-flight messages", it just removes them from the queue when you call `receive_messages`, so `delete_messages` can never work -- indeed if you print out its return value as it is today, `Successful` is always empty, everything's always marked `Failed` 😕)

Reply:
Yeah, that makes sense. These two points — that we're mimicking SQS and that this toy implementation is only used in tests — are the large pieces I was missing. 🙈